### DISCUSSION TITLE
Nornir, Netbox and connection variables

### DISCUSSION THREAD

**Discussion by ohai89** [nornir-automation/nornir#818] [Q&A] [answered]
Hi,

I can't seem to figure out how to pass in variables to Nornir tasks (for devices polled from Netbox).
In this case I need to set up a custom ssh-port, and only way I have so far been able to do this, is by stating rather globally:
`nr.inventory.defaults.port = 11111`

The main parts of the code are as follows:

````
nr = InitNornir(
    runner={"plugin": "threaded", "options": {"num_workers": 25}},
    inventory={
        "plugin": "NetBoxInventory2",
        "options": {
            "nb_url": NETBOX_URL,
            "nb_token": NETBOX_TOKEN,
            "hosts": {
                "port": args.port,
                "connection_options": {
                    "scrapli": {"extras": {"hostkey_verify": False}},
                },
            },
        },
    },
)
devices = nr.filter(name=args.host)
result = devices.run(task=netmiko_send_command, command_string="/user ssh-keys print")
print_result(result)
````

I tried around with scrapli, that's why the connection options are there.
I'm quite certain this is not particularly hard thing to do, just haven't figured out how to do it.
All the examples I have been able to find are for SimpleInventory which specifies these in a separate config.yml.

I would want to be able to inject the port, and maybe other variables as well based on other script logic. The `inventory.defaults` will not cut it as I have different ports for different host platforms..

**Comment by ohai89** [comment, marked as answer]
It seems that by manually setting the `task.host.open_connection()` I can pass in arbitrary variables for the connection:

```
def open_connection(
        self,
        connection: str,
        configuration: Config,
        hostname: Optional[str] = None,
        username: Optional[str] = None,
        password: Optional[str] = None,
        port: Optional[int] = None,
        platform: Optional[str] = None,
        extras: Optional[Dict[str, Any]] = None,
        default_to_host_attributes: bool = True,
    ) -> ConnectionPlugin:
```

So I came up with:
```
def nornir_task(task: Task) -> Result:
    task.host.open_connection(
        connection="netmiko", configuration=task.nornir.config, port=args.port
    )
    result = task.run(task=netmiko_send_command, command_string="/user ssh-keys print")
    task.host.close_connection(connection="netmiko")
    return result

result = devices.run(task=nornir_task)
```

I wonder if there are prettier ways of accomplishing this?

**Comment by dbarrosop** [comment]
You can read about `transform_functions` in the docs, it is usually meant for small tweaks to the inventory.

**Reply by ohai89** [reply]
Hi, thanks, this looks a lot finer way of doing what I am trying to do here.
However I am running to:
```
Exception has occurred: PluginNotRegistered
plugin <function adapt_host_data at 0x7f5c0fa54700> is not registered
```

```
def adapt_host_data(host, port):
    host.port = port

inventory_options = {"port": args.port}
nr = InitNornir(
    runner={"plugin": "threaded", "options": {"num_workers": 50}},
    inventory={
        "plugin": "NetBoxInventory2",
        "options": {"nb_url": NETBOX_URL, "nb_token": NETBOX_TOKEN},
        "transform_function": adapt_host_data,
        "transform_function_options": inventory_options,
    },
)
```
This is pretty much copy & paste from v2.5.0 how-to.
I'm running Nornir v3.3.0, if that has anything to do with what I see here.

**Reply by dbarrosop** [reply]
Please, read: https://nornir.readthedocs.io/en/latest/plugins/index.html#registering-plugins
